tests/e2e/test_spyre_cb.py

```diff
@@ -83,177 +91,412 @@ def test_cb_handling(
+    engine_core = EngineCore(vllm_config=vllm_config,
+                             executor_class=executor_class,
+                             log_stats=False)
+    scheduler: ContinuousBatchingSpyreScheduler = engine_core.scheduler
+
+    # Create random requests of specified lengths and max_tokens
+    sorted_reqs_params = zip(steps_add_reqs, seqs_max_tokens, prompts_lengths)
+    requests: deque[tuple[int, EngineCoreRequest]] = deque()
+    for i, (add_step, max_tokens,
+            prompt_length) in enumerate(sorted_reqs_params):
+        # ignoring eos because we want to force the decoding to finish
+        # after max_tokens exactly
+        sampling_params = SamplingParams(max_tokens=max_tokens,
+                                         temperature=0.0,
+                                         ignore_eos=True)
+        request = create_random_request(request_id=i,
+                                        num_tokens=prompt_length,
+                                        sampling_params=sampling_params)
+        requests.append((add_step, request))
+
+    # Run steps, until last step is reached (provided through parameter)
+    request_outputs = []
+    for step in range(last_step):
+        # Add requests for this step
+        while requests and requests[0][0] == step:
+            engine_core.add_request(requests.popleft()[1])
+        
+        # If initial step: check that tkv is zero and all requests are waiting
+        if step == 0:
+            assert scheduler.tkv == 0, "Step 0 (init), tkv"
+            assert len(scheduler.waiting) != 0, "Step 0 (init), num waiting"
+            assert len(scheduler.running) == 0, "Step 0 (init), num running"
+
+        # Check step if it is in the provided list of steps to check
+        if checked_steps and step == checked_steps[0]["step"]:
+            step_ref = checked_steps.popleft()
+            reqs_output_ids = [
+                req_output.request_id for req_output in request_outputs
+            ]
+            
+            assert scheduler.tkv == step_ref["tkv"], \
+                f"Step {step}, tkv"
+            assert len(scheduler.waiting) == step_ref["num_waiting"], \
+                f"Step {step}, num waiting"
+            assert len(scheduler.running) == step_ref["num_running"], \
+                f"Step {step}, num running"
+            assert reqs_output_ids == step_ref["request_outputs"], \
+                f"Step {step}, request outputs"
+
+        # Perform next step
+        request_outputs = engine_core.step().outputs
+
+    # Assert there is no more running or waiting request after last step
+    assert len(scheduler.waiting) == 0, \
+        "Last step done but still there are waiting requests"
+    assert len(scheduler.running) == 0, \
+        "Last step done but still there are running requests"
+
+    # Tkv in scheduler is cleared one step later
+    _ = engine_core.step().outputs
+    assert scheduler.tkv == 0, "Tkv not cleared two steps after last step"
```

Comment:
Honestly we can even include a step 74 that checks for everything cleared
```
        {
            "step": 74,
            "tkv": 0,
            "num_waiting": 0,
            "num_running": 0,
            "request_outputs": [],
        },
```

this can eliminate all these lines, although it does take a hit on readability. Will leave it up to you

Reply:
mmh
maybe the looper can have no idea of how it should finishes, neither about the initialization and just test everything given the elements in `checked_steps`. That's probably the simplest.